agent/usage_report.go

```diff
@@ -0,0 +1,116 @@
+package agent
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+var (
+	errNoData     = errors.New("no data to report")
+	jsonMarshaler = &pmetric.JSONMarshaler{}
+)
+
+// usageTracker is a store for usage data. It keeps track of the last usage data and the delta between each update.
+type usageTracker struct {
+	// lastUsageData is a map of the last cumulative usage data for each signal.
+	lastUsageData map[usageSignal]usage
+
+	mut               sync.Mutex
+	lastDataPoints    []usage
+	currentDataPoints []usage
+}
+
+func newUsageTracker() *usageTracker {
+	return &usageTracker{
+		lastUsageData:     make(map[usageSignal]usage),
+		currentDataPoints: make([]usage, 0),
+		lastDataPoints:    make([]usage, 0),
+	}
+}
+
+// Add records the current cumulative usage and calculates the delta between the last update.
+func (ur *usageTracker) Add(data usage) {
+	ur.mut.Lock()
+	defer ur.mut.Unlock()
+
+	if data.val == 0 {
+		return
+	}
+
+	deltaTraceUsage := data.val - ur.lastUsageData[data.signal].val
```

Comment:
What happens if `ur.lastUsageData[data.signal].val` is 0 or if `ur.lastUsageData[data.signal]` is nil?

Reply:
When a new signal shows up for the first time, the `ur.lastUsageData[data.signal]` will return the zero value for `usage` type. The zero value of `usage` is 
```
usage{
    signal:    "", // Zero value for string
    val:       0,  // Zero value for float64
    timestamp: time.Time{}, // Zero value for time.Time
}
```
So when the lone `deltaTraceUsage := data.val -ur.lastUsageData[data.sginal].val` executes, it's effectively computing `data.val - 0` for a new signal that hasn't been seen before